File: cpp/029.cpp
class Solution {
public:
    int divide(int dividend, int divisor) {
        if (divisor == 0 || divisor == -1 && dividend == INT_MIN)
            return INT_MAX;
        int sign = ((dividend > 0) ^ (divisor > 0)) ? -1 : 1;
        long long divAbs = abs((long long)dividend), diAbs = abs((long long)divisor);
        if (diAbs == 1)
            return divAbs * sign;
        int ret = 0;
        while (divAbs >= diAbs) {
            long long temp = diAbs;
            int shift = 1;
            while (divAbs >= (temp << 1)) {
                temp <<= 1;
                shift <<= 1;
            }
            ret += shift;
            divAbs -= temp;
        }
        return ret * sign;
    }
};

File: go/029.go
func divide(dividend int, divisor int) int {
    //overflow
    if dividend == math.MinInt32 && divisor == -1 {
        return math.MaxInt32
    }
    if divisor == 1 {
        return dividend
    }
    if divisor == -1 {
        return -dividend
    }
    sign := 1
    if dividend > 0 && divisor < 0 || dividend < 0 && divisor > 0 {
        sign = -1
    }
    
    divAbs, dvAbs, ret := abs(int64(dividend)), abs(int64(divisor)), 0
    for divAbs >= dvAbs {
        temp, shift := dvAbs, 1
        for divAbs >= (temp << 1) {
            temp <<= 1
            shift <<= 1
        }
        divAbs -= temp
        ret += shift

    }
    return ret * sign
}

func abs(a int64) int64 {
    if a < 0 {
        return -a
    }
    return a
}
